lib/external_service_new/shopify_downloader.rb

```diff
@@ -28,37 +29,122 @@ def each_email(id:, &block)
 
       page = 1
       loop do
-        response = dispatcher.dispatch(:get, "/admin/customers/search.json?page=#{page}&limit=250&fields=email")
+        response = dispatcher.dispatch(:get, "/admin/customers/search.json?page=#{page}&limit=250&fields=#{EMAIL_FIELDS}&query= " + parse_filters(@remote_filters))
         break if response["customers"].empty?
 
-        response["customers"].each do |customer|
-          next unless customer && customer["email"]
+        response["customers"].each do |shopify_customer|
+          next unless matching_after_remote_filters(shopify_customer)
 
-          yield customer["email"]
+          yield shopify_customer["email"]
         end
         page += 1
       end
 
       logger && logger.info("Downloaded Shopify Customer List")
     end
 
+
     private
 
-    attr_reader :dispatcher, :filters, :logger
+    def parse_filters(filters)
+      filters.map{|f| "#{f.fetch(:code)}:#{f.fetch(:value)}"}.join(" ")
+    end
+
+    def matching_after_remote_filters(shopify_customer)
+      shopify_customer["total_spent"] = shopify_customer["total_spent"].to_i
+      shopify_customer["orders_count"] = shopify_customer["orders_count"].to_i
+      shopify_customer["created_at"] = shopify_customer["created_at"].to_time
+      shopify_customer["updated_at"] = shopify_customer["updated_at"].to_time
+
+      @after_remote_filters.all? do |filter|
+        self.send("check_#{filter.fetch(:code)}", shopify_customer, filter.fetch(:value))
+      end
+    end
+
+
+    # AFTER_REMOTE_FILTER CHECKS
+
+    def check_created_at_max(shopify_customer, time)
+      shopify_customer["created_at"] <= time
+    end
+
+    def check_created_at_min(shopify_customer, time)
+      shopify_customer["created_at"] >= time
+    end
+
+    def check_updated_at_max(shopify_customer, time)
+      shopify_customer["updated_at"] <= time
+    end
+
+    def check_updated_at_min(shopify_customer, time)
+      shopify_customer["updated_at"] >= time
+    end
+
+    def check_average_basket_revenue_max(shopify_customer, value)
+      return false if shopify_customer["orders_count"] < 1
+
+      (shopify_customer["total_spent"] / shopify_customer["orders_count"]) <= value.to_i
+    end
+
+    def check_average_basket_revenue_min(shopify_customer, value)
+      return false if shopify_customer["orders_count"] < 1
+
+      (shopify_customer["total_spent"] / shopify_customer["orders_count"]) >= value.to_i
+    end
+
+    def check_orders_count_max(shopify_customer, value)
+      shopify_customer["orders_count"] <= value.to_i
+    end
+
+    def check_orders_count_min(shopify_customer, value)
+      shopify_customer["orders_count"] >= value.to_i
+    end
+
+    def check_total_spent_max(shopify_customer, value)
+      shopify_customer["total_spent"] <= value.to_i
+    end
+
+    def check_total_spent_min(shopify_customer, value)
+      shopify_customer["total_spent"] >= value.to_i
```

Comment:
I feel the methods `.to_i` and `to_time` should be part of some sort of parameter validation.